src/librustdoc/clean/types.rs

```diff
@@ -388,6 +388,42 @@ impl Item {
         self.def_id.as_def_id().map(|did| tcx.get_attrs(did).inner_docs()).unwrap_or(false)
     }
 
+    crate fn visibility(&self, tcx: TyCtxt<'_>) -> Visibility {
+        let mut def_id = match self.def_id {
+            // Anything but DefId *shouldn't* matter, but return a reasonable value anyway.
+            ItemId::Auto { .. } | ItemId::Blanket { .. } => return Visibility::Inherited,
+            ItemId::Primitive(..) => return Visibility::Public,
+            ItemId::DefId(did) => did,
+        };
+        match *self.kind {
+            // Variant fields inherit their enum's visibility.
+            StructFieldItem(..) if is_field_vis_inherited(tcx, def_id) => {
+                return Visibility::Inherited;
+            }
+            // Variants always inherit visibility
+            VariantItem(..) => return Visibility::Inherited,
+            // Return the visibility of `extern crate` statement, not the crate itself (which will always be public).
+            ExternCrateItem { crate_stmt_id, .. } => def_id = crate_stmt_id.to_def_id(),
+            // Trait items inherit the trait's visibility
+            AssocConstItem(..) | AssocTypeItem(..) | TyMethodItem(..) | MethodItem(..) => {
+                let is_trait_item = match tcx.associated_item(def_id).container {
+                    ty::TraitContainer(_) => true,
+                    ty::ImplContainer(impl_id) => tcx.impl_trait_ref(impl_id).is_some(),
+                };
+                if is_trait_item {
+                    return Visibility::Inherited;
+                }
+            }
+            _ => {}
+        }
+        // If this item was inlined, show the visibility of the `use` statement, not the definition.
+        let def_id = match self.inline_stmt_id {
+            Some(inlined) => inlined,
+            None => def_id,
+        };
```

Comment:
Shouldn't this check be done before the big `match *self.kind {` above? It seems like this does extra work for inlined items, and the code's a bit harder to understand.

Reply:
No, because then all the uses of `def_id` in the match would be wrong (e.g. `associated_item` and overriding the `ExternCrateItem` def_id).

Reply:
also I'm not sure what you mean by extra work - this isn't doing an early return, it's *just* reassigning `def_id`.

Reply:
What I'm suggesting is changing this to be an early return. What's the point of calculating `def_id` earlier if this code overrides it anyway?